cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,420 @@
+			expectedListResult:     newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+			expectedUpdateResponse: newFakeNotFoundResponse(),
+			expectedUpdateError:    errors.New("failed to create firewall after found missing on update"),
+			expectedError:          errors.New("failed to create firewall after found missing on update"),
+		},
+		{
+			name:                   "return error when failing to add inbound rules on update request to firewall API",
+			fwCache:                newFakeFirewallCache(fakeWorkerFWName, diffFakeInboundRule),
+			inboundRules:           []godo.InboundRule{fakeInboundRule},
+			expectedListResponse:   newFakeOKResponse(),
+			expectedListResult:     newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+			expectedUpdateResponse: newFakeOKResponse(),
+			expectedUpdateError:    errors.New("failed to add new firewall inbound rules"),
+			expectedError:          errors.New("failed to add new firewall inbound rules"),
+		},
+		{
+			name:                 "when the firewall cache is nil, return existing firewall from API and update cache",
+			fwCache:              newFakeFirewallCacheEmpty(),
+			inboundRules:         []godo.InboundRule{fakeInboundRule},
+			expectedListResponse: newFakeOKResponse(),
+			expectedListResult:   newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+			expectedGetResponse:  newFakeOKResponse(),
+			expectedGetResult:    newFakeFirewall(fakeWorkerFWName, fakeInboundRule),
+		},
+	}
+
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			gclient := newFakeGodoClient(
+				&fakeFirewallService{
+					listFunc: func(context.Context, *godo.ListOptions) ([]godo.Firewall, *godo.Response, error) {
+						return []godo.Firewall{*test.expectedListResult}, test.expectedListResponse, test.expectedListError
+					},
+					addRulesFunc: func(ctx context.Context, fID string, rr *godo.FirewallRulesRequest) (*godo.Response, error) {
+						return test.expectedAddRulesResponse, test.expectedAddRulesError
+					},
+					updateFunc: func(ctx context.Context, fID string, fr *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+						return test.expectedUpdateResult, test.expectedUpdateResponse, test.expectedUpdateError
+					},
+					createFunc: func(ctx context.Context, fr *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+						return test.expectedCreateResult, test.expectedCreateResponse, test.expectedCreateError
+					},
+					getFunc: func(ctx context.Context, fID string) (*godo.Firewall, *godo.Response, error) {
+						return test.expectedGetResult, test.expectedGetResponse, test.expectedGetError
+					},
+				},
+			)
+			fwManagerOp = newFakeFirewallManagerOp(gclient, test.fwCache)
+			fc := NewFirewallController(ctx, kclient, gclient, inf.Core().V1().Services(), fwManagerOp, []string{})
+
+			err := fc.fwManager.Set(ctx, test.inboundRules)
+			if want, got := test.expectedError, err; !reflect.DeepEqual(want, got) {
```

Comment:
Error can be compared directly by value again.

Reply:
addressed in comment above